samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java

```diff
@@ -0,0 +1,295 @@
+        StreamEdge streamEdge = jobGraph.getOrCreateEdge(streamSpec);
+        GraphNode node = getOrCreateStreamNode(streamEdge);
+        buildJsonGraph(stream, node, opNodes, outputToStreamSpecs);
+      });
+
+    Job job = new Job();
+    job.setJobName(jobNode.getJobName());
+    job.setJobId(jobNode.getJobId());
+    job.setOpNodes(createListAndSort(opNodes.values()));
+    return job;
+  }
+
+  private void buildJsonGraph(MessageStream inputStream, GraphNode node, Map<Integer, OpNode> opNodes, Map<OutputStream, StreamEdge> outputToStreamSpecs) {
+    Collection<OperatorSpec> specs = ((MessageStreamImpl) inputStream).getRegisteredOperatorSpecs();
+    for (OperatorSpec spec : specs) {
+      GraphNode next = getOrCreateOpNode(opNodes, spec);
+      node.getSuccessors().add(next.getNodeId());
+
+      if (spec instanceof SinkOperatorSpec) {
+        StreamEdge streamEdge = outputToStreamSpecs.get(((SinkOperatorSpec) spec).getOutStream());
+        GraphNode streamNode = getOrCreateStreamNode(streamEdge);
+        next.getSuccessors().add(streamNode.getNodeId());
+        next = streamNode;
+      }
+
+      if (spec.getNextStream() != null) {
+        buildJsonGraph(spec.getNextStream(), next, opNodes, outputToStreamSpecs);
+      }
+    }
+  }
+
+  private OpNode getOrCreateOpNode(Map<Integer, OpNode> opNodes, OperatorSpec opSpec) {
+    OpNode node = opNodes.get(opSpec.getOpId());
+    if (node == null) {
+      node = new OpNode();
+      node.setNodeId(nextID++);
+      node.setOpCode(opSpec.getOpCode().name());
+      node.setOpId(opSpec.getOpId());
+      opNodes.put(opSpec.getOpId(), node);
+    }
+
+    if (opSpec instanceof PartialJoinOperatorSpec) {
+      // every join will have two partial join operators
+      // we will choose one of them in order to consolidate the inputs
+      // the first one who registered with the outputStreamToJoinSpec will win
+      MessageStream output = opSpec.getNextStream();
+      OperatorSpec joinSpec = outputStreamToJoinSpec.get(output);
+      if (joinSpec == null) {
+        joinSpec = opSpec;
+        outputStreamToJoinSpec.put(output, joinSpec);
+      } else {
+        OpNode joinNode = opNodes.get(joinSpec.getOpId());
+        joinNode.setPairNodeId(node.getNodeId());
+        node.setPairNodeId(joinNode.getNodeId());
+      }
+    }
+
+    return node;
+  }
+
+  private StreamNode getOrCreateStreamNode(StreamEdge edge) {
+    String streamId = edge.getStreamSpec().getId();
+    StreamNode node = streamNodes.get(streamId);
+    if (node == null) {
+      node = new StreamNode();
+      node.setNodeId(nextID++);
+      node.setStreamId(streamId);
+      node.setSystem(edge.getStreamSpec().getSystemName());
+      node.setPhysicalName(edge.getStreamSpec().getPhysicalName());
+      node.setPartition(edge.getPartitionCount());
+      streamNodes.put(streamId, node);
+    }
+    return node;
+  }
+
+  private static <T extends GraphNode> List<T> createListAndSort(Collection<T> nodes) {
```

Comment:
Minor: Name makes it sound like a more general helper than it is. Maybe `sortGraphNodesById`?